src/beast/beast/hash/hash_append.h

```diff
@@ -21,125 +21,165 @@
 template <class T>
-struct is_contiguously_hashable
-    : public std::integral_constant <bool,
-        std::is_integral<T>::value || 
-        std::is_enum<T>::value     ||
-        std::is_pointer<T>::value>
+/*constexpr*/
+inline
+void
+maybe_reverse_bytes(T& t, std::false_type)
 {
-};
+    reverse_bytes(t);
+}
+
+template <class T, class Hasher>
+/*constexpr*/
+inline
+void
+maybe_reverse_bytes(T& t, Hasher&)
+{
+    maybe_reverse_bytes(t, std::integral_constant<bool,
+        Hasher::endian == endian::native>{});
+}
 
-// If this fails, something is wrong with the trait
-static_assert (is_contiguously_hashable<int>::value, "");
+} // detail
+
+// is_uniquely_represented<T>
+
+// A type T is contiguously hashable if for all combinations of two values of
+// 	a type, say x and y, if x == y, then it must also be true that
+// 	memcmp(addressof(x), addressof(y), sizeof(T)) == 0. I.e. if x == y,
+// 	then x and y have the same bit pattern representation.
+
+template <class T>
+struct is_uniquely_represented
+    : public std::integral_constant<bool, std::is_integral<T>::value ||
+                                          std::is_enum<T>::value     ||
+                                          std::is_pointer<T>::value>
+{};
+
+template <class T>
+struct is_uniquely_represented<T const>
+    : public is_uniquely_represented<T>
+{};
+
+template <class T>
+struct is_uniquely_represented<T volatile>
+    : public is_uniquely_represented<T>
+{};
+
+template <class T>
+struct is_uniquely_represented<T const volatile>
+    : public is_uniquely_represented<T>
+{};
+
+// is_uniquely_represented<std::pair<T, U>>
 
-// pair
 template <class T, class U>
-struct is_contiguously_hashable <std::pair<T, U>>
-    : public std::integral_constant <bool,
-        is_contiguously_hashable<T>::value && 
-        is_contiguously_hashable<U>::value &&
-        sizeof(T) + sizeof(U) == sizeof(std::pair<T, U>)>
+struct is_uniquely_represented<std::pair<T, U>>
+    : public std::integral_constant<bool, is_uniquely_represented<T>::value && 
+                                          is_uniquely_represented<U>::value &&
+                                          sizeof(T) + sizeof(U) == sizeof(std::pair<T, U>)>
 {
 };
 
-#if ! BEAST_NO_IS_CONTIGUOUS_HASHABLE_PAIR
-static_assert (is_contiguously_hashable <std::pair <
-    unsigned long long, long long>>::value, "");
-#endif
+// is_uniquely_represented<std::tuple<T...>>
 
-#if ! BEAST_NO_TUPLE_VARIADICS
-// std::tuple
 template <class ...T>
-struct is_contiguously_hashable <std::tuple<T...>>
-    : public std::integral_constant <bool,
-        static_and <is_contiguously_hashable <T>::value...>::value && 
-        static_sum <sizeof(T)...>::value == sizeof(std::tuple<T...>)>
+struct is_uniquely_represented<std::tuple<T...>>
+    : public std::integral_constant<bool,
+            static_and<is_uniquely_represented<T>::value...>::value && 
+            static_sum<sizeof(T)...>::value == sizeof(std::tuple<T...>)>
 {
 };
```

Comment:
Nitpick: Inconsistent formatting. Some of the `struct is_uniquely_represented` have the braces on one line. These have the braces on two lines. I don't care, but you might.